client/src/worldgen.rs

```diff
@@ -40,83 +46,551 @@ impl NodeState {
+                                _ => (Material::Sand, -1.0),
                             };
+
+                            // maximum max_elevation for this voxel according to the max_elevations
+                            // of the incident nodes that dictate the content of this chunk
+                            let max_e = trilerp(&max_elevations, p) - elevation_boost;
+
+                            if self.surface.voxel_elevation(q, cube) < max_e / -10.0 {
+                                voxels.data_mut()[index(p)] = voxel_mat;
+                            }
                         }
                     }
                 }
             }
-            _ if self.solid_material().is_some() => {
+            _ if self.kind.solid_material().is_some() => {
                 for z in GAP..(SUB - GAP) {
                     for y in GAP..(SUB - GAP) {
                         for x in GAP..(SUB - GAP) {
-                            let i = subchunk_index(x, y, z, subchunk_offset);
-                            voxels.data_mut()[i] = self.solid_material().unwrap();
+                            let i = index(absolute_subchunk_coords(x, y, z, subchunk_offset));
+                            voxels.data_mut()[i] = self.kind.solid_material().unwrap();
                         }
                     }
                 }
             }
             _ => unreachable!(),
         }
     }
+}
 
-    pub fn solid_material(self) -> Option<Material> {
-        match self {
-            Hole(_) | RootSky | Sky | DeepSky => Some(Material::Void),
-            DeepLand => Some(Material::Stone),
-            _ => None,
+pub fn voxels(graph: &mut DualGraph, node: NodeId, cube: Vertex) -> VoxelData {
+    let max_elevations = chunk_incident_max_elevations(graph, node, cube);
+
+    let mut voxels = VoxelData::Uninitialized;
+
+    for ([x, y, z], path) in cube.dual_vertices() {
+        let state = graph
+            .get(path.fold(node, |node, side| graph.neighbor(node, side).unwrap()))
+            .as_ref()
+            .unwrap();
+        let subchunk_offset = na::Vector3::new(x as usize, y as usize, z as usize);
+        state.write_subchunk(&mut voxels, subchunk_offset, cube, max_elevations);
+    }
+
+    voxels
+}
+
+fn chunk_incident_max_elevations(graph: &DualGraph, node: NodeId, cube: Vertex) -> [f64; 8] {
+    let mut e = cube
+        .dual_vertices()
+        .map(|(_, path)| path.fold(node, |node, side| graph.neighbor(node, side).unwrap()))
+        .map(|n| graph.get(n).as_ref().unwrap().max_elevation as f64);
+
+    // this is a bit cursed, but I don't want to collect into a vec because perf,
+    // and I can't just return an iterator because then something still references graph.
+    let e1 = e.next().unwrap();
+    let e5 = e.next().unwrap();
+    let e3 = e.next().unwrap();
+    let e7 = e.next().unwrap();
+    let e2 = e.next().unwrap();
+    let e6 = e.next().unwrap();
+    let e4 = e.next().unwrap();
+    let e8 = e.next().unwrap();
```

Comment:
Could we modify `trilerp` to take arguments in the same order `dual_vertices` yields them, or vis versa, and thereby avoid this shuffle?